src/wallet/rpc/wallet.cpp

```diff
@@ -582,6 +582,124 @@ static RPCHelpMan upgradewallet()
+            }
+        },
+        RPCExamples{
+            HelpExampleCli("simulaterawtransaction", "[\"myhex\"]")
+            + HelpExampleRpc("simulaterawtransaction", "[\"myhex\"]")
+        },
+    [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
+{
+    std::shared_ptr<const CWallet> const wallet = GetWalletForJSONRPCRequest(request);
+    if (!wallet) return NullUniValue;
+    const CWallet* pwallet = wallet.get();
+
+    RPCTypeCheck(request.params, {UniValue::VARR, UniValue::VOBJ}, true);
+
+    LOCK(pwallet->cs_wallet);
+
+    CAmount changes{0};
+    std::set<COutPoint> spent;
+
+    UniValue include_watchonly(UniValue::VNULL);
+    if (request.params[1].isObject()) {
+        UniValue options = request.params[1];
+        RPCTypeCheckObj(options,
+            {
+                {"include_watchonly", UniValueType(UniValue::VBOOL)},
+            },
+            true, true);
+
+        include_watchonly = options["include_watchonly"];
+    }
+
+    isminefilter filter = ISMINE_SPENDABLE;
+    if (ParseIncludeWatchonly(include_watchonly, *pwallet)) {
+        filter |= ISMINE_WATCH_ONLY;
+    }
+
+    const auto& txs = request.params[0].get_array();
+    std::map<COutPoint, CAmount> new_utxos; // UTXO:s that were made available in transaction array
+    for (size_t i = 0; i < txs.size(); ++i) {
+        CMutableTransaction mtx;
+        if (!DecodeHexTx(mtx, txs[i].get_str(), /* try_no_witness */ true, /* try_witness */ true)) {
+            throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Transaction hex string decoding failure.");
+        }
+
+        // Fetch previous transactions (inputs)
+        std::map<COutPoint, Coin> coins;
+        for (const CTxIn& txin : mtx.vin) {
+            coins[txin.prevout]; // Create empty map entry keyed by prevout.
+        }
+        pwallet->chain().findCoins(coins);
+
+        // Fetch debit; we are *spending* these; if the transaction is signed and
+        // broadcast, we will lose everything in these
+        for (const auto& txin : mtx.vin) {
+            auto& op = txin.prevout;
+            if (spent.count(op)) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Transaction(s) are spending the same output more than once");
+            }
+            if (new_utxos.count(op)) {
+                changes -= new_utxos.at(op);
+                new_utxos.erase(op);
+            } else {
+                if (!coins.count(op)) {
+                    throw JSONRPCError(RPC_INVALID_PARAMETER, "One or more transaction inputs are missing");
+                }
+                if (coins.at(txin.prevout).IsSpent()) {
+                    throw JSONRPCError(RPC_INVALID_PARAMETER, "One or more transaction inputs have been spent already");
+                }
+                changes -= pwallet->GetDebit(txin, filter);
+            }
+            spent.insert(txin.prevout);
+        }
+
+        for (size_t i = 0; i < txs.size(); ++i) {
```

Comment:
In 84a745eed211b597a89f3bbc3da73b0ac2cb457d "rpc/wallet: add simulaterawtransaction RPC"

This loop doesn't make sense here. It shouldn't be necessary to iterate `txs` at the same time we are already iterating `txs`.

At best, this could be moved outside of the first `txs` loop as a preprocessing step to extract all of the UTXOs created, but I don't think that would be a good idea. That would allow for transactions that have circular dependencies and those would be invalid. It is fine to have an order requirement as that is what package validation requires.

Reply:
Yeah, this was clearly a mistake. I removed the loop and kept the inner block only.